### PR TITLE
Feature/exchange credential

### PR DESCRIPTION
**Pull request by jakubkoci**
This implements the second sub-task of the issue #64 + threads which were necessary for passing attachments around. 

Some other notes worth mentioning:
* There are no revocations yet, I added some dummy code just to create a credential.
* `master_secret` is always generated as a new one and it's not reused. I think that would need to be addressed in #65 
* I introduced `JsonEncoder` utility class which eventually could and should be used everywhere we encode/decode.

It's finally ready to review, now.

### CODE REVIEW

File: src/lib/wallet/IndyWallet.ts [hunk 1]
```diff
@@ -76,6 +76,52 @@ export class IndyWallet implements Wallet {
...
+    credDef: CredDef,
+    masterSecretName: string
+  ): Promise<[CredReq, CredReqMetadata]> {
+    if (!this.wh) {
+      throw Error('Wallet has not been initialized yet');
+    }
+    // TODO save `masterSecret` during wallet init and just use it in `proverCreateCredentialReq`
+    const masterSecretId = await this.indy.proverCreateMasterSecret(this.wh, masterSecretName);
+    return this.indy.proverCreateCredentialReq(this.wh, proverDid, offer, credDef, masterSecretId);
+  }
```
Comment (TimoGlastra) [inline src/lib/wallet/IndyWallet.ts:91]: Does this mean that credentials cannot be linked at the moment as all credentials use a new `masterSecret`?

Comment (TimoGlastra) [reply src/lib/wallet/IndyWallet.ts:91]: I didn't read your comment in the PR.

File: src/lib/protocols/credentials/messages/CredentialResponseMessage.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import { Expose } from 'class-transformer';
+import { Equals, IsString } from 'class-validator';
+import { AgentMessage } from '../../../agent/AgentMessage';
+import { MessageType } from './MessageType';
+import { Attachment } from './Attachment';
+
+interface CredentialResponseMessageOptions {
+  id?: string;
+  comment: string;
```
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/messages/CredentialResponseMessage.ts]: Updated

File: src/lib/protocols/credentials/messages/CredentialResponseMessage.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+  @Equals(CredentialResponseMessage.type)
+  public readonly type = CredentialResponseMessage.type;
+  public static readonly type = MessageType.CredentialResponse;
+
+  @IsString()
+  public comment!: string;
+
+  @Expose({ name: 'credentials~attach' })
+  public attachments!: Attachment[];
```
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/messages/CredentialResponseMessage.ts:33]: I haven't found that `{ each: true }` parameter in the docs, so I just added `@Type(() => Attachment)`

File: src/lib/protocols/credentials/messages/CredentialRequestMessage.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  @Equals(CredentialRequestMessage.type)
+  public readonly type = CredentialRequestMessage.type;
+  public static readonly type = MessageType.CredentialRequest;
+
+  @IsString()
+  public comment!: string;
+
+  @Expose({ name: 'requests~attach' })
```
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/messages/CredentialRequestMessage.ts:31]: I haven't found that { each: true } parameter in the docs, so I just added @Type(() => Attachment)

File: src/lib/protocols/credentials/messages/CredentialRequestMessage.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import { Equals, IsString } from 'class-validator';
+import { AgentMessage } from '../../../agent/AgentMessage';
+import { MessageType } from '../../credentials/messages/MessageType';
+import { Expose } from 'class-transformer';
+import { Attachment } from './Attachment';
+
+interface CredentialRequestMessageOptions {
+  id?: string;
+  comment: string;
```
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/messages/CredentialRequestMessage.ts]: Updated

File: src/lib/protocols/credentials/CredentialService.ts [hunk 6]
```diff
@@ -21,46 +30,225 @@ export class CredentialService extends EventEmitter {
...
+      mimeType: 'application/json',
+      data: {
+        base64: JsonEncoder.encode(credReq),
+      },
+    });
+    const credentialRequest = new CredentialRequestMessage({
+      comment: 'some credential request comment',
+      attachments: [attachment],
+    });
+    credentialRequest.setThread(new ThreadDecorator({ threadId: credential.tags.threadId }));
```
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/CredentialService.ts]: Aha, that's good to know 👍  Updated

File: src/lib/protocols/credentials/CredentialUtils.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+      };
+    }, {});
+  }
+
+  private static encode(value: any) {
+    if (!isNaN(value)) {
+      return value.toString();
+    }
+
+    return new BigNumber(sha256.array(value)).toString();
```
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/CredentialUtils.ts:33]: Good point. I checked npm that there are no other dependencies for both `js-sha256` and `bn.js`. 

Although `js-sha256` source code contains `Buffer` and `ArrayBuffer` which could be problematic 
 there is a solved issue describing it should work https://github.com/emn178/js-sha256/issues/11

I'll check both tomorrow in our project.

File: src/lib/protocols/credentials/__tests__/CredentialService.test.ts [hunk 8]
```diff
@@ -1,33 +1,145 @@
...
+      '1891661791592401364793544973569850112519453874155294114300886230795255714579603892516573573155105241417827172655027285062713792077137917614458690245067502490043126222829248919183676387904671567784621260696991226361605344734978904242726352512061421137336169348863177667958333777571812458318894495425085637370715152338807798447174855274779220884193480392221426666786386198680359381546692118689959879385498358879593493608080913336396532253364578927496868954362997951935977034507467171417802640352406191044080192001188762610962085274270807255753335099171457405366335155255038768918649029766176047384127483587155470131765852176320591348954350985301805080951657475246349277435569952922829946940821962356900415616036024524136',
+    m_caps: {
+      master_secret:
+        '32296179824587808657350024608644011637567680645343910724911461554002267640642014452361757388185386803499726200537448417105380225841945137943648126052207146380258164316458003146028',
+    },
+    r_caps: {},
+  },
+  nonce: '784158051402761459123237',
+};
+
```
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/__tests__/CredentialService.test.ts]: Let's try that out. I extracted it into `fixtures.ts` file.

File: src/lib/protocols/credentials/messages/CredentialOfferMessage.ts [hunk 9]
```diff
@@ -1,45 +1,70 @@
...
+      this.attributes = options.attributes;
+    }
   }
 
+  @Expose({ name: '@type' })
   @Equals(CredentialPreview.type)
   public readonly type = CredentialPreview.type;
   public static readonly type = MessageType.CredentialPreview;
 
   @Expose({ name: 'attributes' })
```
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/messages/CredentialOfferMessage.ts]: That make sense. Updated.

File: src/lib/protocols/credentials/messages/CredentialOfferMessage.ts [hunk 10]
```diff
@@ -1,45 +1,70 @@
...
+  public attributes!: CredentialPreviewAttribute[];
+
+  public toJSON(): Record<string, unknown> {
+    return classToPlain(this);
+  }
+}
+
+interface CredentialPreviewAttributeOptions {
+  name: string;
+  mimeType: string;
```
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/messages/CredentialOfferMessage.ts]: I updated that for Credential Preview. It's mandatory in other message objects, right?

File: src/lib/protocols/credentials/messages/Attachment.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  mimeType: string;
+  data: {
+    base64: string;
+  };
+}
+
+export class Attachment {
+  public constructor(options: AttachmentOptions) {
+    if (options) {
+      this.id = options.id;
```
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/messages/Attachment.ts]: Updated.

File: src/lib/protocols/credentials/CredentialService.ts [hunk 12]
```diff
@@ -21,46 +30,225 @@ export class CredentialService extends EventEmitter {
...
+    }
+
+    const credentialId = await this.wallet.storeCredential(
+      uuid(),
+      credential.requestMetadata,
+      cred,
+      credentialDefinition
+    );
+
+    credential.credentialId = credentialId;
```
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/CredentialService.ts:222]: As we discussed at today's call I would suggest keeping it as `CredentialRecord` but rename `credential` to `credentialRecord` across the framework to express it represents just a container/envelope around actual credential data stored in a wallet.

File: src/lib/protocols/credentials/CredentialUtils.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/CredentialUtils.ts:33]: I've just tested and it works with RN 0.61.5 like a charm :)

File: src/lib/protocols/credentials/CredentialService.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/protocols/credentials/CredentialService.ts:221]: We'll do the change in a separate PR.

**Comment by jakubkoci** [comment]
Thanks, @TimoGlastra for suggestions. I've addressed almost all of them. Could you check it, please? If you approve PR don't merge it yet. I'll squash commits and select a few of them into a commit message note.

**Comment by TimoGlastra** [comment]
Looking good 😀

**Review by TimoGlastra — APPROVED** [review]

**Pull request merged**